src/java/org/apache/cassandra/dht/NormalizedRanges.java

```diff
@@ -0,0 +1,295 @@
+                    aRange = aIter.hasNext() ? aIter.next() : null;
+                else
+                    aRange = new Range(bRange.right, aRange.right);
+            }
+        }
+
+        while (aRange != null)
+        {
+            remaining.add(aRange);
+            aRange = aIter.hasNext() ? aIter.next() : null;
+        }
+
+        NormalizedRanges<T> result = normalizedRanges(remaining);
+        if (Range.EXPENSIVE_CHECKS)
+            checkState(result.equals(Range.normalize(Range.subtract(this, b))));
+        return result;
+    }
+
+    @VisibleForTesting
+    public NormalizedRanges<T> invert()
+    {
+        if (isEmpty())
+            return this;
+
+        List<Range<T>> result = new ArrayList<>(size() + 2);
+        T minValue = get(0).left.minValue();
+        T left = minValue;
+        for (Range<T> r : this)
+        {
+            if (!r.left.equals(left))
+            {
+                result.add(new Range<>(left, r.left));
+            }
+            left = r.right;
+        }
+
+        // Loop doesn't add the range to the right of the last one
+        Range<T> last = get(size() - 1);
+        if (!last.right.isMinimum())
+            result.add(new Range<>(last.right, minValue));
+
+        result = Range.normalize(result);
+        if (Range.EXPENSIVE_CHECKS)
+            checkState(result.equals(Range.normalize(Range.subtract(ImmutableList.of(new Range<>(minValue, minValue)), this))));
+        return new NormalizedRanges<>(result);
+    }
+
+    public NormalizedRanges<T> intersection(NormalizedRanges<T> b)
+    {
+        if (this.size() == 1 && this.get(0).isFull())
+            return b;
+        if (b.size() == 1 && b.get(0).isFull())
+            return this;
+
+        List<Range<T>> merged = new ArrayList<>();
+        PeekingIterator<Range<T>> aIter = Iterators.peekingIterator(this.iterator());
+        PeekingIterator<Range<T>> bIter = Iterators.peekingIterator(b.iterator());
+        while (aIter.hasNext() && bIter.hasNext())
+        {
+            Range<T> aRange = aIter.peek();
+            Range<T> bRange = bIter.peek();
+
+            int cmp = compareNormalized(aRange, bRange);
+            if (aRange.intersects(bRange))
+            {
+                merged.addAll(aRange.intersectionWith(bRange));
+                if (cmp == 0)
+                {
+                    aIter.next();
+                    bIter.next();
+                }
+                else if (cmp < 0)
+                {
+                    aIter.next();
+                }
+                else
+                {
+                    bIter.next();
+                }
+            }
+            else
+            {
+                if (cmp <= 0)
+                    aIter.next();
+                if (cmp >= 0)
+                    bIter.next();
+            }
+        }
+
+        NormalizedRanges<T> result = normalizedRanges(merged);
+
+        if (Range.EXPENSIVE_CHECKS)
```

Comment:
`Invariants.paranoia(SUPERLINEAR)`?

Reply:
Cassandra properties are supported to be in `CassandraRelevantProperties`?